turbopack/crates/turbopack-core/src/module_graph/async_module_info.rs

```diff
@@ -56,50 +56,53 @@ async fn compute_async_module_info_single(
     let graph = graph.await?;
 
     let self_async_modules = graph
-        .iter_nodes()
-        .map(async |node| Ok((node, *node.is_self_async().await?)))
-        .try_join()
-        .await?
-        .into_iter()
-        .flat_map(|(k, v)| v.then_some(k))
-        .chain(parent_async_modules.iter().copied())
-        .collect::<FxHashSet<_>>();
+        .enumerate_nodes()
+        .map(async |(_, node)| {
+            Ok(match node {
+                super::SingleModuleGraphNode::Module(node) => {
+                    node.is_self_async().await?.then_some(*node)
+                }
+                super::SingleModuleGraphNode::VisitedModule { idx: _, module } => {
+                    // If a module is async in the parent then we need to mark reverse dependencies
+                    // async in this graph as well.
+                    parent_async_modules.contains(module).then_some(*module)
+                }
+            })
+        })
+        .try_flat_join()
+        .await?;
 
     // To determine which modules are async, we need to propagate the self-async flag to all
-    // importers, which is done using a postorder traversal of the graph.
-    //
-    // This however doesn't cover cycles of async modules, which are handled by determining all
-    // strongly-connected components, and then marking all the whole SCC as async if one of the
-    // modules in the SCC is async.
+    // importers, which is done using a reverse traversal over the graph
+    // Because we walks edges in the reverse direction we can trivially handle things like cycles
+    // without actually computing them.
+    let mut async_modules = FxHashSet::default();
 
-    let mut async_modules = self_async_modules;
     let graph_ref = graph.read();
-    graph_ref.traverse_edges_from_entries_dfs(
-        graph.entry_modules(),
+    graph_ref.traverse_edges_from_entries_dfs_reversed(
+        self_async_modules,
         &mut (),
-        |_, _, _| Ok(GraphTraversalAction::Continue),
-        |parent_info, module, _| {
-            let Some((parent_module, ref_data)) = parent_info else {
-                // An entry module
-                return Ok(());
-            };
-
-            if ref_data.chunking_type.is_inherit_async() && async_modules.contains(&module) {
-                async_modules.insert(parent_module);
-            }
-            Ok(())
+        |parent, module, _state| {
```

Comment:
should the variable names `parent` and `module` be switched here?
parent in reverse order is the child actually?